File: algebra_testbed/algebra/physical/select_hkey_ordered.py
#
# Copyright (C) 2011 Akiban Technologies Inc.
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License, version 3,
# as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see http://www.gnu.org/licenses.
#

import physicaloperator

SimpleOperator = physicaloperator.SimpleOperator

class Select(SimpleOperator):

    def __init__(self, input, predicate_rowtype, predicate):
        SimpleOperator.__init__(self, input)
        self._predicate_rowtype = predicate_rowtype
        self._predicate = predicate
        self._selected = False
        self._examined_row = None

    def handle_row(self, row):
        output_row = None
        if row.rowtype is self._predicate_rowtype:
            self._selected = self._predicate(row)
            self._examined_row = row
            if self._selected:
                output_row = row
        elif self._predicate_rowtype.ancestor_of(row.rowtype):
            if self._examined_row and self._examined_row.ancestor_of(row):
                # row is not an orphan
                if self._selected:
                    output_row = row
            else:
                # row is an orphan
                self._selected = False
                self._examined_row = None
        else: # row.rowtype is ancestor or unrelated
            output_row = row
        return output_row

File: src/main/java/com/akiban/qp/physicaloperator/Select_HKeyOrdered.java
/**
 * Copyright (C) 2011 Akiban Technologies Inc.
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses.
 */

package com.akiban.qp.physicaloperator;

import com.akiban.qp.expression.Expression;
import com.akiban.qp.row.Row;
import com.akiban.qp.row.RowHolder;
import com.akiban.qp.rowtype.RowType;

import java.util.ArrayList;
import java.util.List;

class Select_HKeyOrdered extends PhysicalOperator
{
    // Object interface

    @Override
    public String toString()
    {
        return String.format("%s(%s, %s)", getClass().getSimpleName(), predicateRowType, predicate);
    }

    // PhysicalOperator interface

    @Override
    public Cursor cursor(StoreAdapter adapter, Bindings bindings)
    {
        return new Execution(adapter, inputOperator.cursor(adapter, bindings));
    }

    @Override
    public List<PhysicalOperator> getInputOperators() 
    {
        List<PhysicalOperator> result = new ArrayList<PhysicalOperator>(1);
        result.add(inputOperator);
        return result;
    }

    @Override
    public String describePlan()
    {
        return describePlan(inputOperator);
    }

    // GroupScan_Default interface

    public Select_HKeyOrdered(PhysicalOperator inputOperator, RowType predicateRowType, Expression predicate)
    {
        this.inputOperator = inputOperator;
        this.predicateRowType = predicateRowType;
        this.predicate = predicate;
    }

    // Object state

    private final PhysicalOperator inputOperator;
    private final RowType predicateRowType;
    private final Expression predicate;

    // Inner classes

    private class Execution extends SingleRowCachingCursor
    {
        // Cursor interface

        @Override
        public void open(Bindings bindings)
        {
            input.open(bindings);
        }

        @Override
        public boolean next()
        {
            outputRow(null);
            boolean moreInput = input.next();
            while (outputRow() == null && moreInput) {
                Row inputRow = input.currentRow();
                if (inputRow.rowType() == predicateRowType) {
                    // New row of predicateRowType
                    if ((Boolean) predicate.evaluate(inputRow)) {
                        selectedRow.set(inputRow);
                        outputRow(selectedRow.get());
                    }
                } else if (predicateRowType.ancestorOf(inputRow.rowType())) {
                    // Row's type is a descendent of predicateRowType.
                    if (selectedRow.isNotNull() && selectedRow.get().ancestorOf(inputRow)) {
                        outputRow(inputRow);
                    } else {
                        selectedRow.set(null);
                    }
                } else {
                    outputRow(inputRow);
                }
                if (outputRow() == null) {
                    moreInput = input.next();
                }
            }
            return outputRow() != null;
        }

        @Override
        public void close()
        {
            selectedRow.set(null);
            input.close();
        }

        // Execution interface

        Execution(StoreAdapter adapter, Cursor input)
        {
            this.input = input;
        }

        // Object state

        private final Cursor input;
        // selectedRow is the last input row with type = predicateRowType.
        private final RowHolder<Row> selectedRow = new RowHolder<Row>();
    }
}
